File: SkillPivotAPI/Controllers/StudentsController.c
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using SkillPivotAPI.Data;
using SkillPivotAPI.Models;

namespace SkillPivotAPI.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class StudentsController : ControllerBase
    {
        private readonly ApplicationDbContext _context;

        /// <summary>
        /// Constructor to inject the Database Context.
        /// </summary>
        public StudentsController(ApplicationDbContext context)
        {
            _context = context;
        }

        /// <summary>
        /// Retrieves a complete student profile based on the provided UserId.
        /// Includes academic data, demographic info, and verification status.
        /// </summary>
        /// <param name="userId">The unique ID of the user.</param>
        /// <returns>Student object or NotFound if not exists.</returns>
        [HttpGet("user/{userId}")]
        public async Task<ActionResult<Student>> GetStudentByUserId(int userId)
        {
            // Fetching the student record linked to the specific UserId
            var student = await _context.Students.FirstOrDefaultAsync(s => s.UserId == userId);

            if (student == null)
            {
                return NotFound(new { message = "Student profile not found." });
            }

            return Ok(student);
        }

        /// <summary>
        /// Updates the student profile with new information from the client.
        /// Handles Academic, Demographic, and Verification document paths.
        /// </summary>
        /// <param name="userId">The unique ID of the user to update.</param>
        /// <param name="studentData">The updated student data from the request body.</param>
        [HttpPut("user/{userId}")]
        public async Task<IActionResult> UpdateStudentProfile(int userId, [FromBody] Student studentData)
        {
            // Checking if the student record exists in the database
            var existingStudent = await _context.Students.FirstOrDefaultAsync(s => s.UserId == userId);

            if (existingStudent == null)
            {
                return NotFound(new { message = "Student profile not found." });
            }

            // 1. Update Academic Information
            existingStudent.University = studentData.University;
            existingStudent.Degree = studentData.Degree;
            existingStudent.GPA = studentData.GPA;
            existingStudent.Skills = studentData.Skills;

            // 2. Update Personal Demographic Information
            existingStudent.Gender = studentData.Gender;

            // 3. Update Verification and Documents
            // Note: NicDocumentPath will store the file path after a successful upload
            existingStudent.NicDocumentPath = studentData.NicDocumentPath;
            
            // Note: Verification status is typically updated by an Admin, 
            // but we allow updating it if needed here.
            existingStudent.IsVerified = studentData.IsVerified;

            try
            {
                // Committing changes to the SQL Server database
                await _context.SaveChangesAsync();
                return Ok(new { message = "Full profile updated successfully!" });
            }
            catch (Exception ex)
            {
                // Catching database errors such as connection issues or constraint violations
                return StatusCode(500, new { message = "Database update failed: " + ex.Message });
            }
        }
    }
}

File: SkillPivotAPI/Controllers/StudentsController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using SkillPivotAPI.Data;
using SkillPivotAPI.Models;

namespace SkillPivotAPI.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class StudentsController : ControllerBase
    {
        private readonly ApplicationDbContext _context;

        public StudentsController(ApplicationDbContext context)
        {
            _context = context;
        }

        // 1. GET: Student profile by UserId
        [HttpGet("user/{userId}")]
        public async Task<ActionResult<Student>> GetStudentByUserId(int userId)
        {
            // .AsNoTracking() පාවිච්චි කිරීමෙන් performance වැඩි වෙනවා
            var student = await _context.Students.AsNoTracking().FirstOrDefaultAsync(s => s.UserId == userId);

            if (student == null)
            {
                return NotFound(new { message = "Student profile not found." });
            }
            return Ok(student);
        }

        // 2. POST: Create a new student profile
        [HttpPost]
        public async Task<ActionResult<Student>> PostStudent([FromBody] Student student)
        {
            if (student == null) return BadRequest();

            var existingStudent = await _context.Students.AnyAsync(s => s.UserId == student.UserId);
            if (existingStudent)
            {
                return Conflict(new { message = "A student profile already exists for this user." });
            }

            // අගයන් null නම් හිස් string එකක් දාන්න (Manual SQL table එකේ null allow කරලා නැති නිසා)
            student.University ??= "";
            student.Degree ??= "";
            student.GPA ??= "";
            student.Skills ??= "";
            student.Gender ??= "";
            student.NicDocumentPath ??= "";

            _context.Students.Add(student);
            await _context.SaveChangesAsync();

            return CreatedAtAction(nameof(GetStudentByUserId), new { userId = student.UserId }, student);
        }

        // 3. PUT: Update student profile
        [HttpPut("user/{userId}")]
        public async Task<IActionResult> UpdateStudentProfile(int userId, [FromBody] Student studentData)
        {
            if (studentData == null) return BadRequest();

            var existingStudent = await _context.Students.FirstOrDefaultAsync(s => s.UserId == userId);

            if (existingStudent == null)
            {
                return NotFound(new { message = "Student profile not found." });
            }

            // දත්ත Update කිරීම
            existingStudent.University = studentData.University ?? "";
            existingStudent.Degree = studentData.Degree ?? "";
            existingStudent.GPA = studentData.GPA ?? "";
            existingStudent.Skills = studentData.Skills ?? "";
            existingStudent.Gender = studentData.Gender ?? "";
            existingStudent.NicDocumentPath = studentData.NicDocumentPath ?? existingStudent.NicDocumentPath;
            existingStudent.IsVerified = studentData.IsVerified;

            try
            {
                await _context.SaveChangesAsync();
                return Ok(new { message = "Profile updated successfully!" });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error updating database", error = ex.Message });
            }
        }
    }
}